STYLE.md

```diff
@@ -0,0 +1,228 @@
+
+            cleanup:
+              if (NULL != foo)
+              {
+                free(foo);
+              }
+              if (NULL != bar)
+              {
+                free(bar);
+              }
+              if (NULL != baz)
+              {
+                free(baz);
+              }
+              return result;
+            }
+
+
+# Naming Conventions
+- PascalCase for files, functions, types, and structs
+- camelCase for parameters and variables
+- g_ prefix for global variables
+- UPPER_CASE for macros
+
+# Variables
+-  Variables must always be initialized
+-  For C, variables should be defined at the beginning of the code block they're used:
+    - Bad:
+
+            int foo(int bar)
+            {
+              int x=0; // variable defined at the beginning of a larger block, but it's used only in the smaller block
+              if (1 == bar)
+              {
+                x = 7;
+                return x;
+              }
+              int y = 5; // variable defined not at the beginning of the code block
+              return y;
+            }
+
+    - Good:
+
+            int foo(int bar)
+            {
+              int y = 5;
+              if (1 == bar)
+              {
+                int x=0;
+                x = 7;
+                return x;
+              }
+              return y;
+            }
+
+- Variables must not be reused for different purposes, except for simple index variables in for loops and a function-wide result variable
+- Variable names must be verbose enough to explain their purpose if not obvious
+- Use boolean true and false rather than 1 and 0 for boolean operations
+
+# Error Handling
+- For C, return integer status codes for functions that can fail, typically using standard errno values, and 0 for success
+- For C++ using complex/optional error types is allowed
+- Always validate function inputs
+- Use early returns for validation and error conditions
+- Log errors with descriptive messages using OsConfigLogError or OsConfigLogInfo
+- For C++ avoid throwing exceptions, but handle them properly when thrown by external code
+
+# Memory Management
+- Always check for memory allocation failures, fail gracefully if possible
+- Avoid dangling invalid pointers - if it's freed it should be NULL, use FREE_MEMORY(ptr) macro to safely free memory.
+- Free all allocated memory (except output values) before function returns, including error paths
+- Functions allocating memory should document ownership transfer to caller
+- For C++ use smart pointers wherever possible, avoid new and delete
+
+# Logging
+- Include an OsConfigLogHandle log parameter in functions that need to log
+- Use appropriate log levels:
+  - OsConfigLogDebug: Detailed debugging information
+  - OsConfigLogInfo: General operational information
+  - OsConfigLogError: Function errors and failures
+- Log return values and status codes for debugging
+- Avoid logging sensitive data, eg. information read from the user machine
```

Comment:
>Avoid logging sensitive data, eg. information read from the user machine [](http://example.com/codeflow?start=2&length=72)

Unless logging level is debug. At debug level we log everything that is necessary for debugging. Such as, the raw text from system commands (which itself can contain machine or user data)